mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py

```diff
@@ -0,0 +1,209 @@
+            if with_score_factors:
+                score_factors = score_factors[batch_inds, topk_inds, :]
+
+        mlvl_valid_bboxes.append(bbox_pred)
+        mlvl_valid_scores.append(scores)
+        priors = self.anchor_center(priors)  # xyxy2xy by richard.
+        mlvl_valid_priors.append(priors)
+        if with_score_factors:
+            mlvl_score_factors.append(score_factors)
+
+    batch_mlvl_bboxes_pred = torch.cat(mlvl_valid_bboxes, dim=1)
+    batch_scores = torch.cat(mlvl_valid_scores, dim=1)
+    batch_priors = torch.cat(mlvl_valid_priors, dim=1)
+    batch_bboxes = self.bbox_coder.decode(
+        batch_priors, batch_mlvl_bboxes_pred, max_shape=img_shape)
+    if with_score_factors:
+        batch_score_factors = torch.cat(mlvl_score_factors, dim=1)
+
+    if not self.use_sigmoid_cls:
+        batch_scores = batch_scores[..., :self.num_classes]
+
+    if with_score_factors:
+        batch_scores = batch_scores * batch_score_factors
+    if not with_nms:
+        return batch_bboxes, batch_scores
+    post_params = get_post_processing_params(deploy_cfg)
+    max_output_boxes_per_class = post_params.max_output_boxes_per_class
+    iou_threshold = cfg.nms.get('iou_threshold', post_params.iou_threshold)
+    score_threshold = cfg.get('score_thr', post_params.score_threshold)
+    pre_top_k = post_params.pre_top_k
+    keep_top_k = cfg.get('max_per_img', post_params.keep_top_k)
+    return multiclass_nms(
+        batch_bboxes,
+        batch_scores,
+        max_output_boxes_per_class,
+        iou_threshold=iou_threshold,
+        score_threshold=score_threshold,
+        pre_top_k=pre_top_k,
+        keep_top_k=keep_top_k)
+
+
+def integral(x, reg_max=16, project=None):
+    if project is None:
+        project = torch.linspace(0, reg_max, reg_max + 1)
+    project = project.unsqueeze(0).unsqueeze(-1).unsqueeze(-1)
+    print('rewrite mmdet.models.dense_heads.gfl_head.integral by richard')
+    c, h, w = x.shape
+    x = torch.nn.functional.softmax(x.reshape(4, reg_max + 1, h, w), dim=1)
+    out = torch.nn.functional.conv2d(x, project.type_as(x), stride=1, padding=0).squeeze(1)  # c, h, w
+    return out
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    func_name='mmdet.models.dense_heads.gfl_head.'
+              'GFLHead.forward_single')
+def gfl_head__forward_single(ctx,
+                             self,
+                             x,
+                             scale,
+                             **kwargs):
+    """Rewrite `forward_single` of `GFLHead` for default backend.
+    Rewrite this function to deploy model, (by richard.)
+    """
+    cls_feat = x
+    reg_feat = x
+    for cls_conv in self.cls_convs:
+        cls_feat = cls_conv(cls_feat)
+    for reg_conv in self.reg_convs:
+        reg_feat = reg_conv(reg_feat)
+    cls_score = self.gfl_cls(cls_feat)
+    bbox_pred = scale(self.gfl_reg(reg_feat)).float()  # B, C(4*17), H, W
+    batch_preds = []
+    for b_pred in bbox_pred:
```

Comment:
Split a batched tensor with for loop is not a good idea. TensorRT can not support dynamic batch on such structure.

Reply:
And since integral is performed in get_bboxes in mmdetection, I think we can write a batched integral and use it in gfl_head__get_bbox, rewrite of forward is not necessary.